src/orderbook_fetcher.ts

```diff
@@ -0,0 +1,84 @@
+import Web3 from 'web3'
+import { AbiItem } from 'web3-utils'
+import { getOpenOrdersPaginated, Fraction, Orderbook, Offer, OrderBN } from '@gnosis.pm/dex-contracts'
+import { BatchExchangeViewer } from '@gnosis.pm/dex-contracts/build/types/BatchExchangeViewer'
+import batchExchangeViewerAbi from '@gnosis.pm/dex-contracts/build/contracts/BatchExchangeViewer.json'
+import { CategoryLogger } from 'typescript-logging'
+
+interface Network {
+  address: string
+}
+
+export class OrderbookFetcher {
+  orderbooks: Map<string, Orderbook> = new Map()
+  lastUpdatedBatchId = 0
+  batchExchangeViewer: BatchExchangeViewer | null = null
+
+  constructor(readonly web3: Web3, readonly pageSize: number, readonly logger: CategoryLogger) {
+    // Schedule check to update orderbook every 1s
+    setInterval(async () => {
+      await this.maybeUpdateOrderbooks()
```

Comment:
Can this last more than 1s to fetch? what would happen then?

Reply:
This fetch is almost certainly going to take more than 1 seconds. For now, we would exit early in the second invocation as lastUpdatedBatchId will have been updated already.

Given your comment below, I will turn this into a recursive setTimeout invocation and remove the lastBatchId flag.

Reply:
Actually the recursive approach is pretty ugly. I think doing a unconditional update every 30s sounds reasonable. If runs overlap it's not a big issue since state is only written to at the very end of each round, thus there is no data race (we always have the state of the most recently finished query). 

I can make this more explicit by pulling updateOrderbooks out of `this`.

Reply:
ok, then, maybe a basic semaphore flag in a try/finally would be good

Reply:
I don't understand your suggestion. Can you elaborate maybe with a codesnippet? In the current version the only thing in the critical section would be the assignment to `this.orderbooks`.

We can start fetching in parallel without any data assignment races.